### PR TITLE
Add API endpoints for meetings and meeting accesses + refactor user and group accesses

### PR DESCRIPTION
**Pull request by sampaccoud**
## Purpose

We want to be able to retrieve meetings independently. On the list API endpoint, meetings, must be filtered by a datetime range so we know how to compute occurrences for each meeting returned.

## Proposal

Before this, we need to refactor room accesses:
- on the frontend, we want to display user and group accesses in the same list so a simplification is to use the same model to record them. 
- we will need to handle accesses on other resources than rooms, so we split the room model in a resource base model and a room model that inherits from this model.
- meeting accesses don't require roles (a meeting will only ever belong to a user), so we make it a separate MeetingAccess model.

We could then add the new API endpoints for meetings and meeting accesses together with a full test suite to battle test them.

### CODE REVIEW

File: tests/apps/core/test_core_models_resource_accesses.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
+"""
+Unit tests for the RoomAccess model with user
```
Comment (kernicPanel) [inline tests/apps/core/test_core_models_resource_accesses.py:2]:
```suggestion
Unit tests for the ResourceAccess model with user
```

File: tests/apps/core/test_core_models_resource_accesses.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+from magnify.apps.core.factories import (
+    GroupResourceAccessFactory,
+    RoomFactory,
+    UserResourceAccessFactory,
+)
+
+
+class ResourceAccessesModelsTestCase(TestCase):
+    """
+    Unit test suite to validate the behavior of the RoomAccess model with user
```
Comment (kernicPanel) [inline tests/apps/core/test_core_models_resource_accesses.py:16]:
```suggestion
    Unit test suite to validate the behavior of the ResourceAccess model with user
```

File: tests/apps/core/test_core_models_resource_accesses.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        """Room group accesses should be unique."""
+        access = GroupResourceAccessFactory()
+
+        with self.assertRaises(ValidationError) as context:
+            GroupResourceAccessFactory(group=access.group, resource=access.resource)
+
+        self.assertEqual(
+            context.exception.messages,
+            ["Resource access with this group and resource already exists."],
+        )
```
Comment (kernicPanel) [inline tests/apps/core/test_core_models_resource_accesses.py:106]: room -> resource

**Review by kernicPanel — COMMENTED** [review]
Maybe we should split test files that already reach 1000+ lines ?
Something like https://github.com/openfun/marsha/tree/master/src/backend/marsha/core/tests/api/playlists

File: src/magnify/apps/core/models.py [hunk 4]
```diff
@@ -395,7 +422,32 @@ class Meeting(BaseModel):
 
     name = models.CharField(max_length=500)
     room = models.ForeignKey(
-        Room, null=True, blank=True, related_name="meetings", on_delete=models.RESTRICT
+        Room,
+        help_text=_("Host this meeting in a room."),
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply src/magnify/apps/core/models.py:426]: :+1:

File: tests/apps/core/test_core_api_rooms_meetings.py [hunk 5]
```diff
@@ -136,57 +125,260 @@ def test_api_room_meetings_anonymous_public_filter_invalid(self, _mock_token):
...
+            "from=2022-07-7T09:00:00Z&to=2022-07-13T18:00:00Z",
+            HTTP_AUTHORIZATION=f"Bearer {jwt_token}",
+        )
+
+        self.assertEqual(response.status_code, 200)
+        results = response.json()
+        self.assertEqual(len(results), 4)
+        self.assertEqual(
+            {r["id"] for r in results},
+            {str(m.id) for m in meetings},
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply tests/apps/core/test_core_api_rooms_meetings.py:231]: ok then :blush:

File: tests/apps/core/test_core_api_rooms_meetings.py [hunk 6]
```diff
@@ -136,57 +125,260 @@ def test_api_room_meetings_anonymous_public_filter_invalid(self, _mock_token):
...
-        self.assertEqual(response.status_code, 403)
+
+        self.assertEqual(response.status_code, 200)
+        results = response.json()
+        self.assertEqual(len(results), 3)
         self.assertEqual(
-            response.json(),
-            {"detail": "You do not have permission to perform this action."},
+            {r["id"] for r in results},
+            {str(m.id) for m in meetings},
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply tests/apps/core/test_core_api_rooms_meetings.py:279]: yep

[the remark this answers is not in the document]
Comment (sampaccoud) [reply tests/apps/core/test_core_api_rooms_meetings.py:279]: yep

File: tests/apps/core/test_core_api_rooms_meetings.py [hunk 7]
```diff
@@ -136,57 +125,260 @@ def test_api_room_meetings_anonymous_public_filter_invalid(self, _mock_token):
...
             "from=2022-07-7T09:00:00Z&to=2022-07-13T18:00:00Z",
             HTTP_AUTHORIZATION=f"Bearer {jwt_token}",
         )
+
         self.assertEqual(response.status_code, 200)
+        results = response.json()
+        self.assertEqual(len(results), 3)
+        self.assertEqual(
+            {r["id"] for r in results},
+            {str(m.id) for m in meetings},
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply tests/apps/core/test_core_api_rooms_meetings.py:330]: yep

File: tests/apps/core/test_core_api_rooms_meetings.py [hunk 8]
```diff
@@ -136,57 +125,260 @@ def test_api_room_meetings_anonymous_public_filter_invalid(self, _mock_token):
...
+            HTTP_AUTHORIZATION=f"Bearer {jwt_token}",
+        )
+
+        self.assertEqual(response.status_code, 200)
         results = response.json()
         self.assertEqual(len(results), 3)
         self.assertEqual(
             {r["id"] for r in results},
-            {str(meeting_public.id), str(meeting_users.id), str(meeting_groups.id)},
+            {str(m.id) for m in meetings},
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply tests/apps/core/test_core_api_rooms_meetings.py:381]: yep

File: tests/apps/core/test_core_factories_meetings.py [hunk 9]
```diff
@@ -30,17 +26,15 @@ def test_factories_meetings_recurring_until_greater_than_start(self):
...
-        self.assertQuerysetEqual(meeting.users.all(), users, ordered=False)
-
-    def test_factories_meetings_labels(self):
-        """We should be able to attach labels to a meeting."""
-        labels = LabelFactory.create_batch(2)
-        meeting = MeetingFactory(labels=labels)
-        self.assertQuerysetEqual(meeting.labels.all(), labels, ordered=False)
+        MeetingFactory(users=users)
+        self.assertQuerysetEqual(
+            [mgr.user for mgr in MeetingAccess.objects.all()], users, ordered=False
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply tests/apps/core/test_core_factories_meetings.py:31]: let's call it `access`

File: tests/apps/core/test_core_factories_meetings.py [hunk 10]
```diff
@@ -30,17 +26,15 @@ def test_factories_meetings_recurring_until_greater_than_start(self):
...
+        )
 
     def test_factories_meetings_groups(self):
         """We should be able to attach groups to a meeting."""
         groups = GroupFactory.create_batch(2)
-        meeting = MeetingFactory(groups=groups)
-        self.assertQuerysetEqual(meeting.groups.all(), groups, ordered=False)
+        MeetingFactory(groups=groups)
+        self.assertQuerysetEqual(
+            [mgr.group for mgr in MeetingAccess.objects.all()], groups, ordered=False
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply tests/apps/core/test_core_factories_meetings.py:39]: :+1:

File: tests/apps/core/test_core_models_meeting_accesses.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
+"""
+Unit tests for the RoomAccess model with user
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply tests/apps/core/test_core_models_meeting_accesses.py:2]: :+1:

File: tests/apps/core/test_core_models_meeting_accesses.py [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+    def test_models_meeting_accesses_user_str(self):
+        """The str representation should consist in the user name and meeting ID."""
+        access = UserMeetingAccessFactory(user__name="François")
+        self.assertEqual(
+            str(access),
+            f"François is guest in meeting {access.meeting.id!s}",
+        )
+
+    def test_models_meeting_accesses_user_unique(self):
+        """Room user accesses should be unique."""
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply tests/apps/core/test_core_models_meeting_accesses.py:25]: :+1:

File: tests/apps/core/test_core_models_meeting_accesses.py [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+    def test_models_meeting_accesses_group_str(self):
+        """The str representation should consist in the group name and meeting ID."""
+        access = GroupMeetingAccessFactory(group__name="teachers")
+        self.assertEqual(
+            str(access),
+            f"Users in group teachers are guests in meeting {access.meeting_id!s}",
+        )
+
+    def test_models_meeting_accesses_group_unique(self):
+        """Room group accesses should be unique."""
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply tests/apps/core/test_core_models_meeting_accesses.py:45]: :+1:

File: tests/apps/core/test_core_models_resource_accesses.py (hunk 1 above)
Comment (sampaccoud) [reply tests/apps/core/test_core_models_resource_accesses.py:2]: :+1:

File: tests/apps/core/test_core_models_resource_accesses.py (hunk 2 above)
Comment (sampaccoud) [reply tests/apps/core/test_core_models_resource_accesses.py:16]: :+1:

File: tests/apps/core/test_core_models_resource_accesses.py (hunk 3 above)
Comment (sampaccoud) [reply tests/apps/core/test_core_models_resource_accesses.py:106]: good catch!

File: src/magnify/apps/core/api/rooms.py [hunk 14]
```diff
@@ -220,21 +204,21 @@ class RoomGroupAccessViewSet(
...
         """
-        When creating a room group access, ensure that the logged-in user is administrator
+        When creating a room access for a group, ensure that the logged-in user is administrator
         of the related group.
         """
         user = self.request.user
-        group = serializer.validated_data["group"]
-        if not (user and group.administrators.filter(id=user.id).exists()):
+        group = serializer.validated_data.get("group")
+        if group and not (user and group.administrators.filter(id=user.id).exists()):
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply src/magnify/apps/core/api/rooms.py:208]: Ah no. In fact `group` can be null when this is a user access. It's either a user access or a group access.

If it is a group access, then we must be sure that the person who is trying to make an access for this group is the group administrator. In fact, groups are something that you make for yourself. Each person has groups and can use them in his/her rooms/meetings.

**Comment by sampaccoud** [comment]
@kernicPanel I treated your comments. For splitting the test files of more than 1000 lines, it's a good idea that I will do it in another PR!

**Review by kernicPanel — COMMENTED** [review]
Still needing some room cleaning :grin: 
Otherwise LGTM

**Comment by sampaccoud** [comment]
@kernicPanel so sorry :sweat: Room was hard to clean-up!

**Review by kernicPanel — APPROVED** [review]

**Pull request merged**